bindings/per-isolate-data.cc

```diff
@@ -6,10 +6,13 @@ namespace dd {
 
 static std::unordered_map<v8::Isolate*, PerIsolateData*> per_isolate_data_;
 
-PerIsolateData::PerIsolateData(v8::Isolate* isolate) {
+PerIsolateData::PerIsolateData(v8::Isolate* isolate)
+  : isolate_(isolate) {
   per_isolate_data_[isolate] = this;
   node::AddEnvironmentCleanupHook(isolate, [](void* data) {
-    delete static_cast<PerIsolateData*>(data);
+    auto perIsolateData = static_cast<PerIsolateData*>(data);
+    per_isolate_data_.erase(perIsolateData->isolate_);
```

Comment:
Ok, this delete makes 100% sense to me.
But can we not have
- cleanup hook erase here
- cleanup hook on delete in the for where the new is done.

This makes things symmetrical, otherwise this function is tied to the implementation of the For implem

Reply:
Well, it should _only_ be constructed through the `For`. Better actually to just make the regular constructor private. If it's not constructed through the `For` you could end up with entries in the `per_isolate_data_` map getting overwritten. I've made a new commit which puts all the responsibility within the `For` function and makes the constructor private.